### PR TITLE
[Feature] Adds basic support for image content in OpenAI chat routes

### PR DESCRIPTION
**Pull request by fozziethebeat**
This adds piping so that OpenAI Chat Requests can forward input images to be properly processed by LLaVa type models.  This includes a bit of piping, expanded OpenAI request types (copied from openai-python) and some testing.

I've tested this manually by running:

```sh
python3 -m sglang.launch_server --model-path liuhaotian/llava-v1.5-7b --tokenizer-path llava-hf/llava-1.5-7b-hf --chat-template vicuna_v1.1 --port 30000
```

and

```sh
python test/srt/test_openai_server.py
```

Output came out as expected:

```
 Paris. The country is known for its rich history, art, culture, and cuisine. It is also home to the
 United Nations Educational,
 Paris. The country is known for its rich history, art, culture, and cuisine. It is also home to the
 United Nations Educational,
 The capital of France is Paris. (pronounced "Par-ee")
 The image displays a graph with various data points, including a line that goes from 0 to 100.

There are several other lines on
 1. United States - Washington D.C.
2. Canada - Ottawa
3. Mexico - Mexico City
```

### CODE REVIEW

File: python/sglang/srt/conversation.py [hunk 1]
```diff
@@ -341,18 +346,31 @@ def generate_chat_conv(
...
+            if isinstance(message.content, str):
+                conv.append_message(conv.roles[0], message.content)
+            else:
+                real_content = ""
+                for content in message.content:
+                    if content.type == "text":
+                        real_content += content.text
+                    if content.type == "image_url":
+                        real_content += "<image>"
+                        conv.append_image(content.image_url.url)
```
Comment (comaniac) [inline python/sglang/srt/conversation.py]: nit
```suggestion
                    elif content.type == "image_url":
                        real_content += "<image>"
                        conv.append_image(content.image_url.url)
```

File: python/sglang/srt/server.py [hunk 2]
```diff
@@ -197,9 +197,11 @@ async def v1_chat_completions(raw_request: Request):
                 request.messages, tokenize=False, add_generation_prompt=True
             )
             stop = request.stop
+            image_data = None
```
Comment (comaniac) [inline python/sglang/srt/server.py:213]: What happen if `image_data` is None?

File: python/sglang/srt/conversation.py (hunk 1 above)
Comment (fozziethebeat) [reply python/sglang/srt/conversation.py]: done

File: python/sglang/srt/server.py (hunk 2 above)
Comment (fozziethebeat) [reply python/sglang/srt/server.py:213]: I added a comment after verifying `GenerateReqInput` skips any image processing.  Basically nothing happens if `image_data` is `None`.

Comment (comaniac) [reply python/sglang/srt/server.py:213]: Sorry let me rephrase my question a bit. What I meant was when I have a `request.messages` in `dict` type, there are two cases:

1. **chat_template_name is NOT specified**. In this case we will use `apply_chat_template`, and `image_data=None`.
2. **chat_template_name is specified**. In this case we generate conv using chat template, and `image_data=conv.image_data`.

My question is, are these two cases semantically equivalent? Or let's say, if I use the same model with temperature=0, can I get the same responses in both cases?

Comment (fozziethebeat) [reply python/sglang/srt/server.py:213]: When `apply_chat_template` works, they'll be the same, but really they're different.  In my testing, you can't use image input with `apply_chat_template` since it expects the `content` dict entry to always be a string and not the structured OpenAI content types I've added.  When passing in an `image_url` via the openai path, things just go wrong and the prompt comes out odd (also it won't have the `<image>` token).  

I'm not exactly sure how to best handle this, I considered modifying `request.messages` before passing it to `apply_chat_template` to be roughly equivalent to what happens in `generate_chat_conv`.  If we do that, then we could actually always extract `image_data` and then they'd be equivalent.

Comment (comaniac) [reply python/sglang/srt/server.py:213]: Got it. Thanks for the clarification. In this case we should enforce chat template when a request has image data. For example, we return error 503 saying the server is using the chat template from the tokenizer and thus does not support image data; otherwise users would just see a weird response due to odd prompt.

Comment (fozziethebeat) [reply python/sglang/srt/server.py:213]: Related, I wanted to make sure I didn't break the flow where `chat_template == None`.  So I made some tweaks to the openai test suite to make sure the image input doesn't run unless requested.

I also ran everything against `meta-llama/Llama-2-7b-chat-hf` in the server since the HF chat template works.  I tested LLaVa 1.5 however the default HF chat template generates no output, which is what I'd expect given how different it is from Vicuna's template.

Comment (fozziethebeat) [reply python/sglang/srt/server.py:213]: Good idea on the 503.  Done, I now check for the failing request format and return a 503.

**Review by comaniac — APPROVED** [review]
LGTM. I'll leave to others for the change in tokenizer manager.

**Comment by merrymercy** [comment]
@fozziethebeat Thanks!

**Comment by fozziethebeat** [comment]
Thanks for taking the contribution!
